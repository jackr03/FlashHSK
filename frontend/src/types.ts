export interface Flashcard {
  character: string;
  pinyin: string;
  translation: string;
  hskLevel: number;
}
