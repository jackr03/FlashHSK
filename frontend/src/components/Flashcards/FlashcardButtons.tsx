interface FlashcardButtonsProps {
  showPreviousButton: boolean;
  showNextButton: boolean;
  onPreviousClick: () => void;
  onNextClick: () => void;
}

// TODO: Use an icon library instead of text
const FlashcardButtons: React.FC<FlashcardButtonsProps> = ({
  showPreviousButton,
  showNextButton,
  onPreviousClick,
  onNextClick,
}) => {
  return (
    <div className="relative w-full py-4">
      {showPreviousButton && <button
          className="absolute left-0 rounded-xl border-2 bg-white px-4 py-2 text-black shadow-md hover:bg-blue-200 hover:text-white"
          onClick={onPreviousClick}
          data-testid="previous-button"
      >
        Prev
      </button>
      }

      {showNextButton && <button
          className="absolute right-0 rounded-xl border-2 bg-white px-4 py-2 text-black shadow-md hover:bg-blue-200 hover:text-white"
          onClick={onNextClick}
          data-testid="next-button"
      >
        Next
      </button>
      }
    </div>
  );
};

export default FlashcardButtons;
