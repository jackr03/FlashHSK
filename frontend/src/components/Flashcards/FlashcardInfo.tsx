import { Flashcard } from '../../types';

interface FlashcardInfoProps {
  flashcard: Flashcard;
  isFlipped: boolean;
}

const FlashcardInfo: React.FC<FlashcardInfoProps> = ({
  flashcard,
  isFlipped,
}) => {
  const { character, pinyin, translation } = flashcard;

  return (
    <div className="flex h-72 w-60 flex-col items-center justify-center rounded-xl bg-white p-8 shadow-md">
      {isFlipped ? (
        <p className="text-xl" data-testid="translation">{translation}</p>
      ) : (
        <>
          <h2 className="mb-2 text-xl" data-testid="pinyin">{pinyin}</h2>
          <h1 className="mb-5 text-6xl font-bold" data-testid="character">{character}</h1>
        </>
      )}
    </div>
  );
};

export default FlashcardInfo;
