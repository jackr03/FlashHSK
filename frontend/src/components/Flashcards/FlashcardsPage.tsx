import { useEffect, useRef } from 'react';
import LoadingBar, { LoadingBarRef } from 'react-top-loading-bar';
import { ClipLoader } from 'react-spinners';

import useFetchFlashcards from '../../hooks/useFetchFlashcards.ts';

import Flashcards from './Flashcards.tsx';

const FlashcardsPage: React.FC = () => {
  const { flashcards, isLoading } = useFetchFlashcards();
  const loadingBarRef = useRef<LoadingBarRef>(null);

  useEffect(() => {
    if (isLoading) {
      loadingBarRef.current?.continuousStart();
    } else {
      loadingBarRef.current?.complete();
    }
  }, [isLoading]);

  return (
    <>
      <LoadingBar color="red" height={4} ref={loadingBarRef} />

      <div className="flex min-h-screen flex-col items-center justify-center bg-gray-100">
        {isLoading ? (
          <ClipLoader color="red" size={50} data-testid="loading-indicator" />
        ) : (
          <Flashcards flashcards={flashcards} />
        )}
      </div>
    </>
  );
};

export default FlashcardsPage;
