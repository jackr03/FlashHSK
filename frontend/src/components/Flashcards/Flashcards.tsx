import { useCallback, useEffect, useState } from "react";
import { Flashcard } from "../../types";
import FlashcardButtons from "./FlashcardButtons";
import FlashcardInfo from "./FlashcardInfo";

interface FlashcardProps {
  flashcards: Flashcard[];
}

const Flashcards: React.FC<FlashcardProps> = ({ flashcards }) => {
  const [currentCardIndex, setCurrentCardIndex] = useState<number>(0);
  const [currentCardIsFlipped, setCurrentCardIsFlipped] =
    useState<boolean>(false);

  const showPreviousButton = currentCardIndex > 0;
  const showNextButton = currentCardIndex < flashcards.length - 1;

  const handlePreviousClick = useCallback(() => {
    setCurrentCardIndex((currentCardIndex) =>
      Math.max(0, currentCardIndex - 1)
    );
    setCurrentCardIsFlipped(false);
  }, []);

  const handleNextClick = useCallback(() => {
    setCurrentCardIndex((currentCardIndex) =>
      Math.min(flashcards.length - 1, currentCardIndex + 1)
    );
    setCurrentCardIsFlipped(false);
  }, [flashcards.length]);

  const handleCardClick = useCallback(() => {
    setCurrentCardIsFlipped((currentCardIsFlipped) => {
      // If the card is already flipped, move to the next card
      if (currentCardIsFlipped) {
        setCurrentCardIndex((currentCardIndex) =>
          Math.min(flashcards.length - 1, currentCardIndex + 1)
        );
      }

      return !currentCardIsFlipped;
    });
  }, [flashcards]);

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      switch (event.code) {
        case "Space":
          handleCardClick();
          break;
        case "ArrowLeft":
          handlePreviousClick();
          break;
        case "ArrowRight":
          handleNextClick();
          break;
      }
    };

    window.addEventListener("keydown", handleKeyDown);

    return () => {
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, [handleCardClick, handlePreviousClick, handleNextClick]);

  return (
    <section className="flex flex-col items-center justify-center">
      <button onClick={handleCardClick}>
        <FlashcardInfo
          flashcard={flashcards[currentCardIndex]}
          isFlipped={currentCardIsFlipped}
        />
      </button>

      <FlashcardButtons
        showPreviousButton={showPreviousButton}
        showNextButton={showNextButton}
        onPreviousClick={handlePreviousClick}
        onNextClick={handleNextClick}
      />
    </section>
  );
};

export default Flashcards;
