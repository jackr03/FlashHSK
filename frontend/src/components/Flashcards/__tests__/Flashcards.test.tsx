import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { Flashcard } from '../../../types.ts';
import Flashcards from '../Flashcards.tsx';

const mockFlashcards: Flashcard[] = [
  {
    character: '我',
    pinyin: 'wǒ',
    translation: 'I, me',
    hskLevel: 1
  },
  {
    character: '你',
    pinyin: 'nǐ',
    translation: 'you',
    hskLevel: 1
  },
  {
    character: '他',
    pinyin: 'tā',
    translation: 'he, him',
    hskLevel: 1
  }
];

describe('Flashcards', () => {
  beforeEach(() => {
    render(<Flashcards flashcards={mockFlashcards} />);
  });

  it('should render correctly', () => {
    const { asFragment } = render(<Flashcards flashcards={mockFlashcards} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the first flashcard by default', () => {
    expect(screen.getByTestId('character')).toHaveTextContent('我');
    expect(screen.getByTestId('pinyin')).toHaveTextContent('wǒ');
    expect(screen.queryByTestId('translation')).not.toBeInTheDocument();
  });

  it('should not show previous button when on first card', () => {
    expect(screen.queryByTestId('previous-button')).not.toBeInTheDocument();
    expect(screen.getByTestId('next-button')).toBeInTheDocument();
  });

  it('should flip the card then go to next card when clicked', async () => {
    expect(screen.getByTestId('character')).toHaveTextContent('我');

    await userEvent.click(screen.getByTestId('flashcard'));
    expect(screen.getByTestId('translation')).toHaveTextContent('I, me');

    await userEvent.click(screen.getByTestId('flashcard'));
    expect(screen.getByTestId('character')).toHaveTextContent('你');
  });

  it('should flip the card then go to next card when pressing space', async () => {
    expect(screen.getByTestId('character')).toHaveTextContent('我');

    await userEvent.keyboard(' ');
    expect(screen.getByTestId('translation')).toHaveTextContent('I, me');

    await userEvent.keyboard(' ');
    expect(screen.getByTestId('character')).toHaveTextContent('你');
  });

  it('should navigate between cards when pressing arrow keys', async () => {
    expect(screen.getByTestId('character')).toHaveTextContent('我');

    await userEvent.keyboard('{ArrowRight}');
    expect(screen.getByTestId('character')).toHaveTextContent('你');

    await userEvent.keyboard('{ArrowLeft}');
    expect(screen.getByTestId('character')).toHaveTextContent('我');
  });
});