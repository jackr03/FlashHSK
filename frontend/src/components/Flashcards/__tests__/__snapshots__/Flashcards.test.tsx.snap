// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flashcards > should render correctly 1`] = `
<DocumentFragment>
  <section
    class="flex flex-col items-center justify-center"
  >
    <button
      data-testid="flashcard"
    >
      <div
        class="flex flex-col items-center justify-center bg-white p-8 rounded-xl shadow-md w-60 h-72"
      >
        <h2
          class="text-xl mb-2"
          data-testid="pinyin"
        >
          wǒ
        </h2>
        <h1
          class="text-6xl font-bold mb-5"
          data-testid="character"
        >
          我
        </h1>
      </div>
    </button>
    <div
      class="relative w-full py-4"
    >
      <button
        class="absolute right-0 border-2 text-black bg-white rounded-xl shadow-md px-4 py-2 hover:bg-blue-200 hover:text-white"
        data-testid="next-button"
      >
        Next
      </button>
    </div>
  </section>
</DocumentFragment>
`;
